apps/omg/lib/omg/validators/merge_transaction_validator.ex

```diff
@@ -0,0 +1,67 @@
+# Copyright 2019 OmiseGO Pte Ltd
+#
+# Licensed under the Apache License, Version 2.0 (the "License");
+# you may not use this file except in compliance with the License.
+# You may obtain a copy of the License at
+#
+#     http://www.apache.org/licenses/LICENSE-2.0
+#
+# Unless required by applicable law or agreed to in writing, software
+# distributed under the License is distributed on an "AS IS" BASIS,
+# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+# See the License for the specific language governing permissions and
+# limitations under the License.
+
+defmodule OMG.MergeTransactionValidator do
```

Comment:
is this OMG.Validators.MergreTransactionValidator?

Reply:
Well, that depends on the conventions we want to go with :) We started with something like `OMG.Validators.Transaction` in the eWallet but quickly realized how painful it was to have everything called the same - so we switched to `OMG.TransactionValidator` in a sub-folder. Luckily, Elixir isn't Ruby and you're not forced to use encapsulated module when adding a folder. 

I wanted to introduce that and start the conversation, we basically have 3 options:

- `OMG.Validators.MergeTransactionValidator` under `omg/validators/` (it's long and doesn't add much value, but hey, we have aliases so it's still fine)
- `OMG.MergeTransactionValidator` under `omg/` (but it's gonna get crowed)
- `OMG.MergeTransactionValidator` under `omg/validators`, the folder is just a group for us devs, not reflected at the code level. 

/discuss

Reply:
I lean towards `OMG.Validators.MergeTransaction`, a combination of the first and dropping the last validator. With the directory hierarchy, it makes it clear this is a type of validator and also what its goal is, so the secondary `Validator` becomes unnecessary.

Reply:
@achiurizo One more thing I remembered on why we made the switch for the eWallet - when every module is called the same (`OMG.Validators.Transaction`, `OMG.DB.Transaction`, `OMG.Formatters.Transaction`), and you want to use more than one in another module, you will need to either use the intermediary module name or resort to using `as: `:

```
alias OMG.Validators
alias OMG.DB

DB.Transaction.do_something()
Validators.Transaction.do_something()
```

```
alias OMG.Validators.Transaction, as: Validator
alias OMG.DB.Transaction
```

Finally, quick question, am I the only one who dislike having to check the folder path for a file to know what it does? 😅Guess it takes some time getting used to.

<img width="714" alt="Screen Shot 2562-09-24 at 12 17 43" src="https://user-images.githubusercontent.com/2647689/65483124-56575b80-dec5-11e9-8178-eaf208df53fd.png">

Reply:
Question! Are we introducing more `Validators` so that we decided to split it into it's separate namespace?

Reply:
Yes, that's what I think for now. I have the feeling we'll find other validating functions that can come here. If you don't think there isn't going to be more, we can remove it for now (and add it when it happens - but at least we're talking about it now).

Note that if we decide to go with one approach or the other, we should make sure we follow it for all new code and slowly refactor the existing.

PS: I'm working my way through the state for future PG work, and I have to admit the most confusing thing so far is that there are `Core` modules everywhere, and I'm always thinking about the wrong one (because I jump around the code, just read the code for another `Core` :(). That's fine for experienced developers I guess, but increase the learning curve for newcomers, so what are the benefits vs better/more explicit naming?

Reply:
> Core modules everywhere

yeah, ditching the convention of `Something` vs `Something.Core` is on my mind too, this is far too annoying. The direction I'd love to take is to have `Something` hold the logic and `Something.Server` or `SomethingServer` or even `SomethingElseServer` hold whatever the OTP construction is used.

An easier to reach alternative could be `SomethingCore`, just drop the dot, would this help?

Reply:
@T-Dnzt 

> One more thing I remembered on why we made the switch for the eWallet - when every module is called the same (OMG.Validators.Transaction, OMG.DB.Transaction, OMG.Formatters.Transaction), and you want to use more than one in another module, you will need to either use the intermediary module name or resort to using as: :

Fair point. I'm all for sticking with the original first suggestion then:

> OMG.Validators.MergeTransactionValidator under omg/validators/ (it's long and doesn't add much value, but hey, we have aliases so it's still fine)

I personally like having the folder hierarchy dictate code proximity and roles but I think your point is valid when we end up having multiple similar named files 💀

Reply:
Be wary of "when". Because even if we have multiple validators, it doesn't mean they'll be able to live in the same namespace. Transactions do have validators, but stateless or stateful shouldn't be in the same namespace (`OMG.State.Transaction.Validators.<>`... or just `OMG.Validators.<>`. Just a word of caution before we write code that "might" have usage.

Reply:
> yeah, ditching the convention of Something vs Something.Core is on my mind too, this is far too annoying. The direction I'd love to take is to have Something hold the logic and Something.Server or SomethingServer or even SomethingElseServer hold whatever the OTP construction is used.

I think that's a great direction to take. 

> An easier to reach alternative could be SomethingCore, just drop the dot, would this help?

Yeah, that would be super helpful!

> I personally like having the folder hierarchy dictate code proximity and roles but I think your point is valid when we end up having multiple similar named files 

Sounds good, as long as we have an explicit module name, I don't mind intermediary ones reflecting the folder hierarchy.

----

So to summarize, for GenServers and their logic, we'd go with something like:

```
TransactionListenerServer (GenServer)
TransactionListener or TransactionListenerCore (Logic)
```

We follow the folder hierarchy, but also append the role of the module at the end:

```
OMG.Validators.TransactionValidator
```

We don't do create unnecessary folders until we have at least two modules to put in (so I'm going to get rid of the `validators` folder in my PR)

We don't need those changes to happen overnight, we can gradually work on them :)